src/Featurizers/Components/DocumentStatisticsEstimator.h

```diff
@@ -246,31 +299,246 @@ inline DocumentStatisticsAnnotationData::DocumentStatisticsAnnotationData(Freque
+                frequencyVector.begin(),
+                frequencyVector.begin() + static_cast<int>(topKTermsValue),
+                targetValue
+            );
+
+            return std::make_tuple<std::uint32_t, long long>(
+                std::move(targetValue),
+                std::move(remainingTargetValue)
+            );
         }()
+    );
+
+    std::uint32_t targetValue = std::get<0>(targetPack);
+    long long remainingTargetValue = std::get<1>(targetPack);
+
+    FrequencyMap::iterator termFrequencyIter = termFrequency.begin();
+    while (termFrequencyIter != termFrequency.end()) {
+        if (termFrequencyIter->second < targetValue) {
+            termFrequencyIter = termFrequency.erase(termFrequencyIter);
+            continue;
+        } else if (termFrequencyIter->second == targetValue) {
+            if (remainingTargetValue == 0) {
+               termFrequencyIter = termFrequency.erase(termFrequencyIter);
+            } else {
+                ++termFrequencyIter;
+                --remainingTargetValue;
+            }
+        } else {
+            ++termFrequencyIter;
+        }
+    }
+
+    return termFrequency;
+}
+
+inline DocumentStatisticsAnnotationData::FrequencyAndIndexMap MergeTwoMapsWithSameKeys(Details::DocumentStatisticsTrainingOnlyPolicy::FrequencyMap termFrequency,
+                                                                                       Details::DocumentStatisticsTrainingOnlyPolicy::IndexMap termIndex) {
+    DocumentStatisticsAnnotationData::FrequencyAndIndexMap termFrequencyAndIndex;
+    for (auto const & termFrequencyIter : termFrequency) {
+        //assert that termFrequencyIter.first is in termIndex
+        auto termIndexIter = termIndex.find(termFrequencyIter.first);
+        if (termIndexIter == termIndex.end())
+            throw std::invalid_argument("the keys in termFrequency and termIndex do not match");
+
+
+        termFrequencyAndIndex.emplace(std::piecewise_construct,
+                                      std::forward_as_tuple(std::move(termFrequencyIter.first)),
+                                      std::forward_as_tuple(std::move(termFrequencyIter.second), std::move(termIndexIter->second)));
+
+    }
+    return termFrequencyAndIndex;
+}
+
+}
+// ----------------------------------------------------------------------
+// |
+// |  Details::DocumentStatisticsTrainingOnlyPolicy
+// |
+// ----------------------------------------------------------------------
+inline Details::DocumentStatisticsTrainingOnlyPolicy::DocumentStatisticsTrainingOnlyPolicy(
+    StringDecorator decorator,
+    AnalyzerMethod analyzer,
+    std::string regexToken,
+    nonstd::optional<IndexMap> existingVocabulary,
+    nonstd::optional<std::uint32_t> topKTerms,
+    std::float_t minDf,
+    std::float_t maxDf,
+    std::uint32_t ngramRangeMin,
+    std::uint32_t ngramRangeMax
+) :
+    _stringDecoratorFunc(std::move(decorator)),
+    _analyzer(std::move(analyzer)),
```

Comment:
>std::move(analyzer) [](start = 14, length = 19)

Validate

Reply:
the same with decorator

Reply:
Add comments or decorate the name to indicate that it is an optional param.

Reply:
sure